### ISSUE TITLE
[Feature] Multiple Environments

### ISSUE THREAD

**Issue by sharat87** [appsmithorg/appsmith#2836]
## Summary

Applications need to hit different API URLs with varying parameters when run during production and when run as staging. A way globally configure the correct parameters within the application based on the environment would be useful.

<img src="https://front.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_5hgjw)

**Comment by hiphop101** [comment]
I suggest create a global environment setting section for each page.
This global environment variable section can be a JSON configure or anything that allows user to put both credential for UAT and production.
Another solution could be similar to that of retool, for each connection, have a tab to switch between production and UAT.

**Comment by Nikhil-Nandagopal** [comment]
@hiphop101 I'm thinking at a Datasource level makes sense because the variables for Sstaging / Prod will largely be the same across an organization. We can also introduce a generic Store data source where non-database specific values and secrets can be saved.

**Comment by hiphop101** [comment]
@Nikhil-Nandagopal yes, attached a screenshot of retool. I think they also did it at data source level
![image](https://user-images.githubusercontent.com/25315912/106719445-4749f700-663d-11eb-979b-2e9a9b88a60e.png)

**Comment by mayurdhurpate** [comment]
Would love to have this feature soon. Cannot think of any other easy way to swap out databases in AppSmith while keeping the application. 

Some ways in which it helps us devs:
1. We can make dashboards while working on a feature in `dev` / `staging` environment, and those are then production-ready for content/product team automatically.
2. Debugging and replicating issues on dev/staging become much easier.

**Comment by vishaldfreak** [comment]
If possible I'd like this feature prioritised to be implemented sooner. 
In my use case I'd want to fetch data from multiple Shopify stores, It'd be helpful if I can save the API key & App Secret for a particular store somewhere and pass it in requests whenever we make a call to fetch data.

**Comment by ericwooley** [comment]
Retool has actually bungled this IMO. They started off right, but didn't add more than 2 environments...

Here are some other people who are not happy about being limited to two environments. Including a longer rant by me.

https://community.retool.com/t/ability-to-create-more-environments/2048

**Comment by Nikhil-Nandagopal** [comment]
@ericwooley thanks for chiming in! We'll definitely have a more generic implementation here
We're currently building a git sync feature where you'll be able to configure a separate datasource on every branch.

**Comment by raphaeltm** [comment]
Just to add to this discussion, since I've brought up similar but different issues elsewhere previously... I think it would be great to have a hierarchy of special config variables that could get overridden at different levels. For example, it would be great if, like Hasura, you could define server-level environment variables which you could then use in an app in AppSmith. Then, it would be great if you could override those at the organization, and app levels. 

Maybe that's overcomplicating things in some ways, but in others I think it would simplify them. If we just have "configuration variables" as a higher level system, then you wouldn't need to create a separate system specific to datasources for managing environments: servers, organizations, and apps can already be used to manage environments.

This would also help with another problem that we have, which is using weird workarounds to check the url to check if we're in staging or prod to change things in JS Objects to deal with API endpoints. If there was a special "config" object that could be modified at the server/org/app level that was usable in datasource configuration as well as elsewhere in apps, that would be _**so so helpful**_.

**Comment by SidHiremath** [comment]
I was just thinking,  Can we have an Environment file for each environment like env_dev.properties, env_prod.properties and load that when we deploy ?

**Comment by Nikhil-Nandagopal** [comment]
@SidHiremath with Git Import, you will be able to have 2 organizations that have prod / staging datasources and you can easily deploy your application with those. This feature however is to let a developer quickly switch between pod and staging versions of those data sources on any branch.

**Comment by SidHiremath** [comment]
> @SidHiremath with Git Import, you will be able to have 2 organizations that have prod / staging datasources and you can easily deploy your application with those. This feature however is to let a developer quickly switch between pod and staging versions of those data sources on any branch.

It might be good to have the env properties defined in the apppsmith for each environment and while deployment we should override the configurations according to the environment.


Also. as a developper, we should not have access to production system, so it should be part of deployment process to change the configurations and we should have some comand to pass those properties.

**Comment by Nikhil-Nandagopal** [comment]
@SidHiremath we will definitely support an env file that can be uploaded but It may not be from the file system because multiple organizations can be there in a single instance of Appsmith. So we will allow admins to upload an env file that can be read by developers

**Comment by rohan-arthur** [comment]
Hello users, 

I am part of the team that is considering building this feature, and would love to hear about your expectations from `multiple environments`. [Here's](https://calendly.com/rohan-arthur/30min) my calendly link for a short call, looking forward to hearing from you! :) 

Thanks,
Rohan

**Comment by ericwooley** [comment]
@rohan-arthur 

I don't have time for a calendly, but retool just released multiple environments, in case you are unaware, I'd check theirs out as well.

**Comment by vuiets** [comment]
We got a new request for this. 
> User wanted to have deployment stages that work against DataSources (DynamoDB) that are specific to the deployment stage (different AWS accounts)
🔗 [Intercom](https://app.intercom.com/a/inbox/y10e7138/inbox/shared/unassigned/conversation/164629100147933)

**Comment by ZackKnopp** [comment]
I have a use case for this as well.

We have a sales funnel / CRM app with an Appsmith + Supabase stack.

We decided we wanted each company we sign up to have their data segregated, so they have a different Supabase instance.
Each instance has a different api url.

We could use multiple environments to create a dev, prod-company1, prod-company2 ... etc

**Comment by Nikhil-Nandagopal** [comment]
@ZackKnopp it sounds like you want to add environments on the fly which isn't something we're planning to tackle right now. Each environment will need to be configured by a developer in the application.

**Comment by ZackKnopp** [comment]
> @ZackKnopp it sounds like you want to add environments on the fly which isn't something we're planning to tackle right now, each environment will need to be configured by a developer in the application.

I think it would work because we'd configure it manually for any companies we add.

We won't really roll it out to more than 10 companies max, since it's built for internal companies that our small investment company owns. And it's a big change from trying to manage many sales funnels in google sheets which are really easy for users to break.

**Comment by ralfwenske** [comment]
I wasn't aware this topic existed already when asking on Discord. 
As I am not using git at this stage I like simple and obvious solutions (and am grateful that appsmith has lots of those...).

There is already a selector for datasource on each query editor page - I could imagine that a functionality that replaces the  datasource on each query / page might not be overly complicated (hoping to get this feature soon).
 
To cover also apps that use more than one datasource, perhaps a simple array might do (with some extra coding effort):
```
[ 
    {"production": "MyDB", "test":"MyDB-Test"}, 
    {"production":"MyOtherDB", "test":"MyOtherDB-Test"}
 ]
```

**Comment by Nikhil-Nandagopal** [comment]
@ralfwenske we're currently working on this solution. WOuld you be up for a quick chat to have a look at what we're building?

**Comment by ralfwenske** [comment]
Sure Nikhil, chat via email? I am on +10 (Australia) so 8 hours different to you?

**Comment by dgorshkov** [comment]
Hi @Nikhil-Nandagopal, @rohan-arthur,
Is there documentation somewhere about working around this, e.g. via the git sync?

**Comment by Nikhil-Nandagopal** [comment]
@dgorshkov Oops! Forgot to add the link. Here you go 
https://community.appsmith.com/t/is-there-the-ability-to-switch-between-production-and-develop-datasources/702/2?u=dancia

**Comment by dgorshkov** [comment]
@Nikhil-Nandagopal Unfortunately I wasn't able to make that video properly (multiple browsers and video players).

**Comment by sribalajig** [comment]
Posting an update for this feature -

**What is in scope?**

- Ability to configure staging and production environments for datasources
- Ability to switch between staging and production environments while editing the app
- Ability to view the deployed version of the app with staging environment configuration, without impacting the actual deployment
- Ability to define role based access controls for the staging and production environments

**What is out of scope?**

- Ability to add new environments (apart from staging and production)

**Designs**

- [Designs by use case](https://www.figma.com/file/FC5XY2B57Fu41s0a0Wz93g/Multiple-Environments?node-id=1316%3A135778)
- [Interactive prototype](https://www.figma.com/file/FC5XY2B57Fu41s0a0Wz93g/Multiple-Environments?node-id=1335%3A149997)

**Timeline for availability**
January 2023

If you have and questions regarding this feature, please feel free to reach out to me. You can find my calendar here -> https://calendly.com/balaji-gopinath/appsmith-user-interview

**Comment by kshep92** [comment]
@sribalajig Good news, but I'm curious as to why you can't have environments other than staging and production? According to this discussion on the Retool forum, it's quite a useful feature https://community.retool.com/t/ability-to-create-more-environments/2048/39

**Comment by juandl** [comment]
Hi guys, any update on this feature? also, panning to be part of the community edition or enterprise?

**Comment by rubabuddin** [comment]
Also interested in this feature. Lot of weird hacks in place to support this currently.

**Comment by Nikhil-Nandagopal** [comment]
@juandl @rubabuddin we're working on this and should have a stable beta version soon. It's going to be available in the business edition. In the community edition, you can achieve multiple environments using a git repo.
1. Sync the app to a git repository using the production credentials
2. Import the app into a new workspace and configure the staging credentials
3. Now you can make changes to the staging app, commit and merge them in. You can then go to the production app and take a pull whenever you need to deploy the changes to users.

**Comment by dgorshkov** [comment]
@Nikhil-Nandagopal 
> It's going to be available in the business edition
So this will be a paid-only feature?

**Comment by Nikhil-Nandagopal** [comment]
@dgorshkov Yes that is correct. You can reach out for a free trial of our business edition. It's packed with lots of great features and is very affordable for small and large teams alike. 
https://www.appsmith.com/pricing

**Comment by juandl** [comment]
@Nikhil-Nandagopal That's great! I would love to try the business edition, but it sounds more expensive than the actual per-user.  Like retool will be actually much cheaper. we have team members who use the platform for 5hrs a day.. that is like 100 hours a month.. which is like 40 USD based on your model price if I'm not wrong. when retool or other lower code tools are 10 USD per user..

**Comment by Nikhil-Nandagopal** [comment]
@juandl that is true if you only require the multiple environments feature alone in the business edition but we ship with a lot more features like granular access control, SSO, branding, unlimited environments & git repos that are only available in their highest tier which they price at 100-150$/user. Also, we've seen that as these tools scale within your organization, most users aren't using the tools for more than a few hours a week. There are some power users but the majority are light users and billing them all the same drives up a huge bill for your team!

**Comment by juandl** [comment]
> @juandl that is true if you only require the multiple environments feature alone in the business edition but we ship with a lot more features like granular access control, SSO, branding, unlimited environments & git repos that are only available in their highest tier which they price at 100-150$/user. Also, we've seen that as these tools scale within your organization, most users aren't using the tools for more than a few hours a week. There are some power users but the majority are light users and billing them all the same drives up a huge bill for your team!

Fair reply. how long is the trial version? maybe we can get a tried and see 🤔 .. 

A question can we downgrade from business to community edition?  without affecting anything?

**Comment by juandl** [comment]
> @juandl that is true if you only require the multiple environments feature alone in the business edition but we ship with a lot more features like granular access control, SSO, branding, unlimited environments & git repos that are only available in their highest tier which they price at 100-150$/user. Also, we've seen that as these tools scale within your organization, most users aren't using the tools for more than a few hours a week. There are some power users but the majority are light users and billing them all the same drives up a huge bill for your team!

@Nikhil-Nandagopal Fair reply. how long is the trial version? maybe we can get a tried and see 🤔 .. 

A question can we downgrade from business to community edition?  without affecting anything?

**Comment by Nikhil-Nandagopal** [comment]
@juandl we generally do a 30-day trial. You can block some time with our team by filling in this form
https://www.appsmith.com/enterprise
You can upgrade and downgrade without any negative consequences.
https://docs.appsmith.com/getting-started/setup/upgrade-to-business-edition/docker#rollback-optional

**Comment by sribalajig** [comment]
@juandl @dgorshkov we have a preview of the multiple environments feature available for you to try out. Please feel free to setup some time on my [calendar](https://calendly.com/balaji-gopinath/appsmith-user-interview) so that I can walk you through it.

**Comment by demimike** [comment]
We are also interested in testing this out if possible.

**Comment by sribalajig** [comment]
@demimike :  I'm be happy to give you a demo of this on a deployment preview. Please feel free to block some time on my calendar here - https://calendly.com/balaji-gopinath/appsmith-user-interview

**Comment by sribalajig** [comment]
We have a live preview for you to try out -> https://envsmith.appsmith.com/user/login. Please sign up with your existing Appsmith username and kick some tires, let me know what you think. All ears in this last leg of the release, so get your feedback in as quick as possible.

**Comment by mohamedgomran** [comment]
> We have a live preview for you to try out -> https://envsmith.appsmith.com/user/login. Please sign up with your existing Appsmith username and kick some tires, let me know what you think. All ears in this last leg of the release, so get your feedback in as quick as possible.

@sribalajig Thank you, some questions
1- I can't find the env in the preview or deployed app, only while editing the app.
2- How to define datasource for different envs, and then use this datasource based on the selected env?
i.e. a REST api data source = staging.domain.com while selecting staging and same datasource = domain.com while selecting the production env

**Comment by sribalajig** [comment]
@mohamedgomran thanks for checking out the deployment preview! I've also updated a video demonstrating the feature - please do check it out.

> I can't find the env in the preview or deployed app


This is a known issue that the team is working on. Once we fix it, you will be able to see the toggle in view mode and switch between environments. 
In the mean time, if you want to try it out, you can add the query parameter `?environment=staging` or  `?environment=production` in your deployed app's URL.

**Comment by mohamedgomran** [comment]
>@mohamedgomran thanks for checking out the deployment preview! I've also updated a video demonstrating the feature - please do check it out.

@sribalajig Thanks for the prompt response, the video is informative, thank you.

Is there an expected date for releasing this for the business edition?

**Comment by sribalajig** [comment]
@mohamedgomran This is expected to be available for business users by March 20.

**Comment by AlexiaAcevedo** [comment]
@mohamedgomran thanks for pointing this out! Our end-users will need to be able to switch from dev -> staging -> production! A toggle in view mode would be a great solution! Thanks @sribalajig

**Comment by Zajozor** [comment]
> This is expected to be available for business users by March 20.

Is the plan for this to be available only for business and higher users, or will this ever be available in the free edition?
thanks 🙏🏻

**Comment by jwalsh-famkit** [comment]
Please add support for setting different custom header values in APIs for each environment, that seems to be missing in the preview

**Comment by geekyme-fsmk** [comment]
Looking forward to see this feature! I have a need for multiple environments deployment :)

**Comment by sribalajig** [comment]
@AlexiaAcevedo To answer your questions - 
1. Yes, I mean toggling the environments in the deployed app. 
2. For the first release we are going to ship with staging and production. But subsequently, we will give you the ability to add custom environments so you can add as many as you like

**Comment by sribalajig** [comment]
@Zajozor this feature is going to be available for the business edition users.

**Comment by jwalsh-famkit** [comment]
> @jwalsh-famkit do you want to be able to have a different set of headers for staging and production? Because if you want the keys to be the same but with different values, you could use dynamic bindings. Please let me know.

definitely want to be able to have different values for the same keys. I didn't see a way to do this with dynamic bindings at the datasource level. Would also like to be able to control which headers are used in each env.

**Comment by sribalajig** [comment]
Hi all, thanks for all the feedback and interest in the multiple environments deployment preview. I have some news to share regarding the feature release.

**TL;DR**
We have decided to not go forward with the release of this feature because we have identified several user experience issues in our implementation.

**Why?**

At the beginning of the project, we had to choose between two fundamentally different approaches. We picked the approach that you see in the deployment preview because we thought that it would be more extensible in the future. However, looking at the implementation, we realised that we have traded off simplicity and intuitive UX for extensibility. This flow touches datasource connectivity, which is a core flow in Appsmith. We believe that going forward with this implementation will add irreversible technical and UX debt.

**What’s next?**

We know that there has been an incredible amount of interest in this feature from the community. Although this is a tough decision, we really think that this is the right call to make keeping the long term in mind. The team is already working on coming up with updated designs that will fix the UX issues that we have identified. We are also working out the timeline for the availability of the new version. We will come back with an update on the progress we are making with the designs and timelines by 29th March.

**Comment by geekyme-fsmk** [comment]
@sribalajig any updates on this? I'm trying to decide whether I should stick with the [git-based](https://docs.appsmith.com/learning-and-resources/how-to-guides/use-git-to-create-different-environments) approach for managing environments or not.

**Comment by jnmoons** [comment]
Also interested in the update. Any news here?

**Comment by sribalajig** [comment]
@geekyme-fsmk @jnmoons As mentioned here, we are back to the drawing board to fix some issues with our implementation. We will back with timelines shortly.
In the mean time, it's best that you use the Git based approach.

**Comment by kylemantesso** [comment]
This is a huge blocker, the git hack doesn't support APIs or authenticated use cases. Any updates at all on when this might be implemented?

**Comment by dzelemba** [comment]
One tip for others who are looking for workarounds while waiting for this feature that I don't see covered in https://docs.appsmith.com/learning-and-resources/how-to-guides/use-git-to-create-different-environments or the YouTube Video:

Datasource definitions are currently shared across Workspaces so you can leverage that to get some version of multi env support:
  - Use Git to store the configs for your app
  - Create a Prod Workspace and a Staging Workspace and import the app via Git into both
  - Use the exact same Datasource name in both workspace, but switch out the API URL + any auth parameters

Then the workflow that you can use is:
  - Make changes in the Staging App in the Staging Workspace, commit them to Git
  - Open up the Prod app and merge in any changes from Git to deploy

Not sure how well this solves other folks' needs, but it works pretty well for what I was looking for, so sharing here in case others find it useful.

**Comment by sribalajig** [comment]
Update on the progress made so far -  

1. We have designs for the core flows involved in configuring environments, setting up the application context to use a particular environment and configuring granular access control for environments. We're happy to do a demo of the design prototype. Please feel free to setup time on [this](https://calendly.com/balaji-gopinath/appsmith-user-interview) calendar for a walkthrough.
2. Development is in full swing and we expect to be ready for a release by the end of June. We are on track to share a deployment preview with the core functionality working in the first week of June. The deployment preview will give you a sense of how the feature works.

We will keep this ticket posted with regular progress updates. Please tag @sribalajig  for any question you may have.

**Comment by sribalajig** [comment]
@kylemantesso the Git workaround should work for authenticated APIs as well. Could you please let me know details of what is not working for you?

**Comment by hholst80** [comment]
> Update on the progress made so far -
> 
>     1. We have designs for the core flows involved in configuring environments, setting up the application context to use a particular environment and configuring granular access control for environments. We're happy to do a demo of the design prototype. Please feel free to setup time on [this](https://calendly.com/balaji-gopinath/appsmith-user-interview) calendar for a walkthrough.
> 
>     2. Development is in full swing and we expect to be ready for a release by the end of June. We are on track to share a deployment preview with the core functionality working in the first week of June. The deployment preview will give you a sense of how the feature works.
> 
> 
> We will keep this ticket posted with regular progress updates. Please tag @sribalajig for any questions you may have.

We would be happy to have access to the EA of this feature. I will ping you guys on Discord, we have an open thread about deployment environments.

**Comment by sribalajig** [comment]
Four months later, we are back with another preview for Environments in Appsmith. While the core concept of Environments remains unchanged, we have implemented it far simpler this time around.

You don’t have to grapple with environment variables in datasource configurations.
Every datasource gets two configs for Staging and Production.

The core tenets of the feature remain the same as before.

The two environments, Staging and Production, now let you see two sets of data in the same app.
Code your apps during development without worrying it will impact Production data.
When you are happy with how your changes work, deploy to both Staging and Production with one click. Your end-users will see the Production version of your apps by default.
As before, you can connect Git branches to these environments and commit both Staging and Production data from environments to each branch. End-users with View access to the app on each brand will see Production data by default.

All first-party databases and APIs, including GraphQL and authenticated APIs, are covered. Airtable, Twilio, HubSpot, and Google Sheets are out of scope right now. I am looking to understand your use cases for this feature. Please tag me here on any feedback or requests.

Here’s a preview for how it works.

https://github.com/appsmithorg/appsmith/assets/15727421/45d69868-8243-402a-87cf-4e04162fe6fe

Here is a live preview of the environments feature - https://ee-1858.dp.appsmith.com/
Please sign up on this instance and kick some tires. Look forward to hearing your feedback.

@hholst80 FYI as well.

**Comment by hholst80** [comment]
@chomni FYI too!

**Comment by hholst80** [comment]
> 

I hope that we will be able to do environments "East-West" as well, by being able to add any number of environments.  We have production, and test appsmith applications. But we also have different "regions", with different config.

prod-westeu, test-westeu, prod-eastus, test-eastus, 

etc.

**Comment by SelaOZ** [comment]
how can we enable this in our project?
